### DISCUSSION TITLE
Looking for a method to set SnapMode

### DISCUSSION THREAD

**Discussion by saaaaaally** [iTwin/itwinjs-core#7457] [Q&A] [answered]
Looking for a method to set the SnapMode when a tool is executed, does any such method exist? 

![image](https://github.com/user-attachments/assets/1e743907-456f-49b7-a7cc-225b4178bf4b)
![image](https://github.com/user-attachments/assets/05227baa-2e7e-47a8-a9f4-f27cb71effa6)

**Comment by bbastings** [comment]
The active snap mode(s) are a user setting, no tool should be changing what the user has selected.
What exactly is your tool doing that you think it needs to force the snap mode to nearest?

**Comment by MarcBedard8** [comment]
We are working on a Reality Mesh Volume tool. We need that to work to pick point on Reality Data Meshes. In Reality data web viewer, we use to override snap mode at iModelApp startup and only permit "nearest" snap. Now in iTwin Experience, that is not a viable solution. User can select any snap mode, but the tool that only work on reality meshes needs to force that mode.

**Comment by bbastings** [comment]
Sounds to me like your tool shouldn't be enabling AccuSnap...

For normal geometric elements, AccuSnap extracts the element geometry on the backend to compute the snap information. AccuSnap doesn't actually do anything for reality meshes, it just uses the depth point returned by readPixels.

**Reply by bbastings** [reply]
Have a look at RotateViewTool (pickDepthPoint) for an example of a tool that will locate reality data when choosing the center of rotation. It does not enable AccuSnap.

**Reply by MarcBedard8** [reply]
We'll look at it. Thanks

**Reply by saaaaaally** [reply]
I tried adding `IModelApp.accuSnap.enableSnap(false);` to disable snap on the volume tool, I don't think this is going to work for us. Without snap it gives an inaccurate volume value and also if you scroll around the shape doesn't stay put because its not snapped to the reality mesh anymore

https://github.com/user-attachments/assets/37af43a4-dec3-489a-b5e7-497c2601ed1f

**Comment by bbastings** [comment, marked as answer]
> IModelApp.accuSnap.enableSnap(false)

That's not sufficient obviously, as I mentioned you need to get the point from the depth buffer instead. You are not going to be able to set the snap mode because that's not it's intended purpose.

Imagine a user that has set their snap mode(s) to keypoint, then they run your tool and not noticing or expecting their snap mode to be changed try placing a line. Now they may make an error because they don't get the exact endpoint they'd expect from keypoint snap.

The above situation "could" be addressed by your tool saving the current snap mode(s) in onPostInstall and restoring the snap mode(s) in onCleanup...but that's still problematic as the user can change the current snap mode(s) while your tool is running, which will then make your tool not work as expected.

Hopefully you can now appreciate why changing the snap mode(s) for your tool isn't the right approach and you will now actually look at the rotate view tool as I suggested (run the tool, see how it works).

* Also, be aware that you can have multiple snap modes active at once, i.e. keypoint, nearest, and intersection, etc. it's not a single snap mode even though the current ui item only lets you set one at a time.
